trio/_core/_run.py

```diff
@@ -1556,7 +1563,12 @@ def run_impl(runner, async_fn, args):
         # scheduling, then there are other things that will probably need to
         # change too, like the deadlines tie-breaker and the non-deterministic
         # ordering of task._notify_queues.)
-        batch = list(runner.runq)
+        #
+        # We sort tasks before shuffling them so that seeding the Random
+        # instance can make the scheduler deterministic, which is important
+        # for testing and debugging, especially with tools such as Hypothesis,
+        # without giving up the advantages of sets everywhere else.
+        batch = sorted(runner.runq, key=Task.sort_key)
```

Comment:
This is the very inner loop of trio's scheduler, so I'm slightly concerned about adding overhead. Of course the whole `_r.shuffle` thing is also dubious overhead... but maybe some simple measurements wouldn't be amiss? If it's non-trivial then we could hide it behind some flag that hypothesis (or someone) sets...

Reply:
I just don't have enough of a sense for typical workloads to write a good benchmark - all my performance instincts are formed for out-of-memory array workloads or other heavy throughput-dominated stuff.  Very happy to run one and optimize accordingly though!

I don't think a heapq would help us here - to get a deterministic ordering we'd need to either run a n-log-n heapsort (and the `sorted` builtin is almost certainly faster), or else have the tasks stored in lists from creation time in which case we don't need to do anything (but we lose whatever benefits we currently get from sets).

Reply:
We don't really have good workloads either, which makes the whole question of performance a bit vague. But for this purpose I just meant something like a silly microbenchmark with 100 tasks that just reschedule themselves over and over, to get some kind of upper-bound on this. Maybe it doesn't matter at all.